### DISCUSSION TITLE
Consider Lager for observable<>

### DISCUSSION THREAD

**Discussion by arximboldi** [hikoworks/hikogui#392] [Ideas]
This library is very interesting. Sadly it is Windows only so I can't use it, but I will definitelly follow it since there are commercial projects where I could eventually use it.

One thing that has caught my attention is the notion of `observable<>`. If you excuse me the self-promotion, I think cursors in the Lager library could be a more powerful way of propagating changes into and and from widgets: https://sinusoid.es/lager. I talk about examples that could be very applicable for this project in this talk: https://www.youtube.com/watch?v=e2-FRFEx8CA

**Comment by takev** [comment]
Hi, I finally took some time to watch the talk, it was very interesting.

Am I right to understand that when you do a `get`, `set` combination on a `cursor`. Then in reality the `cursor` makes a complete copy of the `state`, makes the modification on the copy, then overwrites the `state` with the modified copy. In this case the `state` object is the full state of the application.

At least this is the theory, for performance reasons some kind of copy-on-write or locking would be needed when working with a state that is very large.

Then a second question, possibly related to the previous one. Can the `state` type be created that instead of holding the data internally, it references external data, possibly through an API (like a database) which already represents the full state of the database through a transaction.

**Reply by arximboldi** [reply]
You are correct, in Lager the state is replaced with a copy of itself, sort of to say. This works because I always use it in combination with Immer, which does the structural sharing (or CoW in a more mutable view of the world): https://github.com/arximboldi/immer

Making the state reference database handles is kind of nasty though. If you are going down that road, I recommend using `lager::store` and communicate to the database via `effect`s (the database is exposed to the effects via `lager::deps`). The in-memory state can still be observed via read-only cursors in that case, but it can only be modified by dispatching actions.

**Reply by takev** [reply]
Cool.

So I am going to think about this a little bit more. And also read those papers you mentioned about lenses.

I have been hurt by using dependencies before, so it is likely I will implement somethnig closer to your idea inside my own library.

_I did make a post earlier in this thread, but it disappeared, or was never posted._ I am looking for composability of `state`, where I can make a standard-widget that is composed of child-widgets (in my case a audio device configuration), that takes a cursor to a standard-sub-`state` and where the standard-sub-`state` type is available to compose with the application's `state`. I have to really think about this.

Audio devices also have a problem with them being in control of their configuration, while at the same time the user should be able to override this. On MacOS for example an audio device could decide to change the sample rate, which is fine and could simply update the `state`, no issues there. However from an application's point of view the sample rate was set by the user and it shouldn't have changed, so the application may need to change it back, possibly at a later point in time (when recording is started).

**Comment by takev** [comment]
I think the system could work well with RCU (read-copy-update). So I have started making an implementation of that.

**Reply by arximboldi** [reply]
Nice, good luck!

**Comment by takev** [comment]
So I got part of it working:

If you look at [line 15 of label_widget.cpp](https://github.com/hikogui/hikogui/blob/139cf30b2db0884e778ac488b4d4365f54961087/src/hikogui/widgets/label_widget.cpp#L15) you will see a cursor name "label" being split into sub-cursors pointing to the member variable "icon" and "text".

The "label" is of type `hi::observer<hi::label>`, `hi::label` in turn has the "icon" and "text" member variables.

If you look at [line 181 of label.hpp](https://github.com/hikogui/hikogui/blob/139cf30b2db0884e778ac488b4d4365f54961087/src/hikogui/label.hpp#L181)
You will see a specialisation of a type_trait `hi::selector<>` which adds the ability for a cursor to extract member variables on a custom type.
